### DISCUSSION TITLE
Is it possible to "lock" context attributes?

### DISCUSSION THREAD

**Discussion by lu-maca** [behave/behave#1136] [Q&A] [answered]
Hi,

Is there a way to raise a warning when trying to overwrite a certain attribute of context? For example:

```python
# environment.py
-----------------

@fixture
def fixture_test(ctx):
    ctx.locked_attribute = SomeImportantClass()
    yield ctx.locked_attribute
    ctx.locked_attribute.clear()

# steps.py
-----------------
@step("This step shall do something")
def do_something(ctx):
    # do something
    ctx.locked_attribute = "foo" # I would like to raise here a warning
    ctx.locked_attribute.method_of_SomeImportantClass() # error is raised by python
    
```
Is this possible? For example, is there a way to raise a `behave.runner.ContextMaskWarning` as said in the documentation [1]?

[1] "If an attempt made by user code to overwrite one of these variables, or indeed by behave to overwite a user-set variable, then a [behave.runner.ContextMaskWarning](https://behave.readthedocs.io/en/stable/api.html?highlight=overwrite#behave.runner.ContextMaskWarning) warning will be raised."

**Comment by jenisys** [comment, marked as answer]
`behave` does not protect `user` context variables. As you have stated, it raises only such a warning if you try to override/hide a `behave` context variable.

Currently, you will need to provide such a special `ValueObject` class, like a `Constant` object that protects itself against assignment (and raises a `ContextMaskWarning` if an assignment occurs).

NOTES:
Normally, you need a `use_or_create_context_param(ctx, name, some_factory_func, *args, **kwargs)` helper function that provides/stores a context variable is none exists.

EXAMPLE:

```python
from behave import given, when, then, step
from behave4me.step_util import use_or_create_context_param

@step(„This step shall do something“)
def do_something(ctx):
    # — NOTE: ctx.locked_attribute is only created/assigned if no such named attribute exists.
    locked_attribute = use_or_create_context_param(ctx, „locked_attribute“, SomeImportantClass)
    locked_attribute.method_of_SomeImportantClass()
```

SEE ALSO:
* https://pypi.org/project/kkconst/
* … (other such projects on https://pypi.org )

**Reply by lu-maca** [reply]
When developing a library that uses behave as a framework, it should be probably a good solution to have some "locked" attribute in `ctx`, without forcing the user to use `use_or_create_context_param(...)` to be sure s/he does not overwrite some important library related `ctx` attributes (or at least to let her/him know that s/he's doing that).

Is this a valid point to you?

**Reply by jenisys** [reply]
Sorry, no.

I basically means that you do not have control of your own code that registers `Context attributes`. Therefore, you want to have a safety net.

HINT:
Use descriptors or something similar to achieve what you desire.

SEE ALSO:
* https://docs.python.org/3/howto/descriptor.html#descriptorhowto

**Comment by jenisys** [comment]
SEE ALSO:
* [Context.use_or_create_param(name, ...)](https://github.com/behave/behave/blob/main/behave/runner.py#L212)
* [Context.use_or_assign_param(name, value)](https://github.com/behave/behave/blob/main/behave/runner.py#L194)
